README.md
## The What

(This is a fork of https://github.com/baalexander/node-xmlrpc for supporting
http compression and keep alive connections.)

The xmlrpc module is a pure JavaScript XML-RPC server and client for node.js.

Pure JavaScript means that the [XML parsing](https://github.com/isaacs/sax-js)
and [XML building](https://github.com/robrighter/node-xml) use pure JavaScript
libraries, so no extra C dependencies or build requirements. The xmlrpc module
can be used as an XML-RPC server, receiving method calls and responding with
method responses, or as an XML-RPC client, making method calls and receiving
method responses, or as both.


## The How

### To Install

```bash
npm install maniajs-xmlrpc
```

### To Use

The file client_server.js in the example directory has a nicely commented
example of using xmlrpc as an XML-RPC server and client (they even talk to each
other!).

A brief example:

```javascript
var xmlrpc = require('xmlrpc')

// Creates an XML-RPC server to listen to XML-RPC method calls
var server = xmlrpc.createServer({ host: 'localhost', port: 9090 })
// Handle methods not found
server.on('NotFound', function(method, params) {
  console.log('Method ' + method + ' does not exist');
})
// Handle method calls by listening for events with the method call name
server.on('anAction', function (err, params, callback) {
  console.log('Method call params for \'anAction\': ' + params)

  // ...perform an action...

  // Send a method response with a value
  callback(null, 'aResult')
})
console.log('XML-RPC server listening on port 9091')

// Waits briefly to give the XML-RPC server time to start up and start
// listening
setTimeout(function () {
  // Creates an XML-RPC client. Passes the host information on where to
  // make the XML-RPC calls.
  var client = xmlrpc.createClient({ host: 'localhost', port: 9090, path: '/'})

  // Sends a method call to the XML-RPC server
  client.methodCall('anAction', ['aParam'], function (error, value) {
    // Results of the method response
    console.log('Method response for \'anAction\': ' + value)
  })

}, 1000)
```

Output from the example:

```
XML-RPC server listening on port 9090
Method call params for 'anAction': aParam
Method response for 'anAction': aResult
```

### Date/Time Formatting

XML-RPC dates are formatted according to ISO 8601. There are a number of
formatting options within the boundaries of the standard. The decoder detects
those formats and parses them automatically, but for encoding dates to ISO
8601 some options can be specified to match your specific implementation.


The formatting options can be set through
```xmlrpc.dateFormatter.setOpts(options);```, where the ```options```
parameter is an object, with the following (optional) boolean members:

* ```colons``` - enables/disables formatting the time portion with a colon as
separator (default: ```true```)
* ```hyphens``` - enables/disables formatting the date portion with a hyphen
as separator (default: ```false```)
* ```local``` - encode as local time instead of UTC (```true``` = local,
```false``` = utc, default: ```true```)
* ```ms``` - enables/disables output of milliseconds (default: ```false```)
* ```offset``` - enables/disables output of UTC offset in case of local time
(default: ```false```)


Default format: 20140101T11:20:00


UTC Example:
```javascript
xmlrpc.dateFormatter.setOpts({
  colons: true
, hyphens: true
, local: false
, ms: true
}) // encoding output: '2014-01-01T16:20:00.000Z'
```

Local date + offset example:
```javascript
xmlrpc.dateFormatter.setOpts({
  colons: true
, hyphens: true
, local: true
, ms: false
, offset: true
}) // encoding output: '2014-01-01T11:20:00-05:00'
```

### Cookies support

It is possible to turn on cookies support for XML-RPC client by special options
flag. If turned on then all the cookies received from server will be bounced
back with subsequent calls to the server. You also may manipulate cookies
manually by the setCookie/getCookie call.

```javascript
var client = xmlrpc.createClient({
  host: 'localhost',
  port: 9090,
  cookies: true
});

client.setCookie('login', 'bilbo');

//This call will send provided cookie to the server
client.methodCall('someAction', [], function(error, value) {
  //Here we may get cookie received from server if we know its name
  console.log(client.getCookie('session'));
});

```

### Custom Types
If you need to parse to a specific format or need to handle custom data types
that are not supported by default, it is possible to extend the serializer
with a user-defined type for your specific needs.

A custom type can be defined as follows:
```javascript
var xmlrpc = require('xmlrpc');
var util = require('util');

// create your custom class
var YourType = function (raw) {
  xmlrpc.CustomType.call(this, raw);
};

// inherit everything
util.inherits(YourType, xmlrpc.CustomType);

// set a custom tagName (defaults to 'customType')
YourType.prototype.tagName = 'yourType';

// optionally, override the serializer
YourType.prototype.serialize = function (xml) {
  var value = somefunction(this.raw);
  return xml.ele(this.tagName).txt(value);
}
```

and then make your method calls, wrapping your variables inside your new type
definition:

```javascript
var client = xmlrpc.createClient('YOUR_ENDPOINT');
client.methodCall('YOUR_METHOD', [new YourType(yourVariable)], yourCallback);
```

### To Debug (client-side)

Error callbacks on the client are enriched with request and response
information and the returned body as long as a http connection was made,
to aide with request debugging. Example:

```javascript
var client = xmlrpc.createClient({ host: 'example.com', port: 80 });
client.methodCall('FAULTY_METHOD', [], function (error, value) {
  if (error) {
    console.log('error:', error);
    console.log('req headers:', error.req && error.req._header);
    console.log('res code:', error.res && error.res.statusCode);
    console.log('res body:', error.body);
  } else {
    console.log('value:', value);
  }
});

// error: [Error: Unknown XML-RPC tag 'TITLE']
// req headers: POST / HTTP/1.1
// User-Agent: NodeJS XML-RPC Client
// ...
// res code: 200
// res body: <!doctype html>
// ...
```

### To Test

[![Build
Status](https://secure.travis-ci.org/baalexander/node-xmlrpc.png)](http://travis-ci.org/baalexander/node-xmlrpc)

XML-RPC must be precise so there are an extensive set of test cases in the test
directory. [Vows](http://vowsjs.org/) is the testing framework and [Travis
CI](http://travis-ci.org/baalexander/node-xmlrpc) is used for Continuous
Integration.

To run the test suite:

`make test`

If submitting a bug fix, please update the appropriate test file too.


## The License (MIT)

Released under the MIT license. See the LICENSE file for the complete wording.


## Contributors

Thank you to all [the
authors](https://github.com/baalexander/node-xmlrpc/graphs/contributors) and
everyone who has filed an issue to help make xmlrpc better.


HISTORY.md
## 2.0.0 / 2016-03-20

 * Http client is now using SuperAgent.
 * Support for GZip/deflate compression.
 * Support for Keep-Alive.

## 1.3.1 / 2015-05-23

 * Adds option to specify xml encoding.
 * Updates XMLBuilder to 2.6.
 * Adds support for Node v0.12.

## 1.3.0 / 2015-01-09

 * Reworked DateFormatter to improve ISO-8601 implementation.
 * Updates XMLBuilder to 2.4.
 * Updates sax to 0.6.
 * Adds more detail to Client request errors to ease debugging.

## 1.2.0 / 2014-01-12

 * Adds (tested) support for Node v0.10 and v0.11.
 * Drops support for Node v0.6.
 * Adds a Custom Type serializer for non-standard XML-RPC types.

## 1.1.1 / 2013-09-22

 * Adds CDATA deserialization.
 * Updates XMLBuilder to get rid of warnings at installation time.
 * Cleans up unformatted code.
 * Removes vestigial code from ancient v0.4 days.

## 1.1.0 / 2012-11-13

 * Supports optional listening handler in Server.
 * Adds close function to server.
 * Adds cookie support to client.
 * Changes minimum Node engine to v0.6.

## 1.0.2 / 2012-07-29

 * Server responds with a 404 if it does not handle the method.
 * Client returns an error if the server returns a 404.

## 1.0.1 / 2012-04-29

 * Fixes content-length header value when sending multi-byte characters.

## 1.0.0 / 2012-03-28

 * Replaces builder/parser logic with marshaller/unmarshaller.
 * Uses XML-RPC's nil for JavaScript null values and vice versa.
 * Client and parser are now re-entrant safe.
 * Moves test strings to fixture files.
 * Adds 30+ test cases.
 * Special thanks to @agnat for his work on the (un)marshaller and tests.

## 0.9.4 / 2012-03-03

 * Handles chunked method calls.
 * Supports ISO-8859-1 encodings.

## 0.9.3 / 2012-01-21

 * Values with no type now default to String as per XML-RPC spec.

## 0.9.2 / 2011-12-10

 * Replaces the XML parser with the Sax.js module.
 * Uses Travic CI for Continuous Integration.
 * Fixes accidental global variable.

## 0.9.1 / 2011-11-29

 * Returns an Error on invalid XML-RPC.
 * Updates to latest version of xmlbuilder to fix install warning.

## 0.9.0 / 2011-11-01

 * Supports the Base64 datatype.
 * Supports the i8 datatype, treating it as a float.
 * Fixes issue where not returning for faults containing no params.
 * Standardizes how to handle empty params. Last failing test now passes.
 * Enforces null value for error param in callbacks.
 * Updates to latest version of xmlbuilder.

## 0.8.1 / 2011-09-01

 * Supports passing the URI as a string to client or server.
 * Host is now an optional parameter for client or server.
 * Fixes bug when performing a method call multiple times.
 * Removes node_modules directory. Use `npm install .` if cloning.

## 0.8.0 / 2011-08-14

 * Supports HTTPS server and client.
 * Improves Basic Auth support.
 * Errors returned are now an instance of Error, not a String.
 * Fixes bug with structs and whitespace.
 * Fixes bug with empty arrays responses.

## 0.7.1 / 2011-08-02

 * Handles chunked method responses.
 * Fixes parsing multi-line strings values in the String parameter.
 * Allows for custom headers in the HTTP request.

## 0.7.0 / 2011-07-12

 * Renames Client.call() to Client.methodCall().
 * Adds better support for sending and parsing empty String parameters.
 * Client handles errors on http request. Includes handling of invalid URLs.
 * Updates documentation.

## 0.6.2 / 2011-06-15

 * Fixes issue with parsing non-value whitespace in method calls.

## 0.6.1 / 2011-06-03

  * Supports CDATA when generating XML calls or responses.

## 0.6.0 / 2011-05-18

  * Initial release to NPM. Considered stable enough for public use.


test/client_test.js
var vows   = require('vows')
  , assert = require('assert')
  , http   = require('http')
  , zlib   = require('zlib')
  , Client = require('../lib/client')
  , fs     = require('fs');

const VALID_RESPONSE = fs.readFileSync(__dirname + '/fixtures/good_food/string_response.xml');
const BROKEN_XML = fs.readFileSync(__dirname + '/fixtures/bad_food/broken_xml.xml');

vows.describe('Client').addBatch({
  //////////////////////////////////////////////////////////////////////
  // Test Constructor functionality
  //////////////////////////////////////////////////////////////////////
  'A constructor' : {
    // Test standard Client initialization
    'with URI options only' : {
      topic: function () {
        var client = new Client({ host: 'localhost', port: 9999, path: '/'}, false)
        return client.options
      }
    , 'contains the standard headers' : function (topic) {
      var headers = {
          'User-Agent': 'NodeJS XML-RPC Client/' + require('./../package').version
        , 'Content-Type': 'text/xml'
        , 'Accept': 'text/xml'
        , 'Accept-Charset': 'UTF8'
        , 'Connection': 'Keep-Alive'
      };
      assert.deepEqual(topic, {
        host: 'localhost',
        port: 9999,
        path: '/',
        url: 'http://localhost:9999/',
        method: 'POST',
        headers: headers,
        keepAlive: 0,
        gzip: false
      });
    }
  }
    // Test passing string URI for options
  , 'with a string URI for options' : {
      topic: function () {
        var client = new Client('http://localhost:9999', false)
        return client.options
      }
    , 'parses the string URI into URI fields' : function (topic) {
        assert.strictEqual(topic.host, 'localhost')
        assert.strictEqual(topic.path, '/')
        assert.equal(topic.port, 9999)
      }
    }
    // Test passing custom headers to the Client
  , 'with options containing header params' : {
      topic: function () {
        var client = new Client({ host: 'localhost', port: 9999, path: '/', headers: { 'User-Agent': 'Testaroo', 'Authorization': 'Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==' }}, false)
        return client.options
      }
    , 'does not overwrite the custom headers' : function (topic) {
        var headers = {
          'User-Agent': 'Testaroo'
        , 'Authorization': 'Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ=='
        , 'Content-Type': 'text/xml'
        , 'Accept': 'text/xml'
        , 'Accept-Charset' : 'UTF8'
        , 'Connection': 'Keep-Alive'
        }
        assert.deepEqual(topic, {
          host: 'localhost',
          url: 'http://localhost:9999/',
          port: 9999,
          path: '/',
          method: 'POST',
          headers: headers,
          keepAlive: 0,
          gzip: false
        });
      }
    }
    // Test passing HTTP Basic authentication credentials
  , 'with basic auth passed' : {
      topic: function () {
        var client = new Client({ basic_auth: { user: 'john', pass: '12345' } }, false)
        return client.options.headers
      }
    , 'correctly encodes and sets the \'Authorization\' header' : function (topic) {
        assert.isNotNull(topic.Authorization)
        assert.equal(topic.Authorization, 'Basic am9objoxMjM0NQ==')
      }
    }
  , 'with a string URI inside options' : {
      topic: function () {
        var client = new Client({url:'http://localhost:9999'}, false)
        return client.options
      }
      , 'parses the string URI into URI fields' : function (topic) {
        assert.strictEqual(topic.host, 'localhost')
        assert.strictEqual(topic.path, '/')
        assert.equal(topic.port, 9999)
      }
    }
    // Test passing encoding
  , 'with an encoding passed': {
      topic: function () {
        var client = new Client({ url:'http://localhost:9999', encoding: 'utf-8' }, false)
        return client.options
      }
    , 'caches the encoding option' : function (topic) {
        assert.strictEqual(topic.encoding, 'utf-8')
      }
    }
  }
  //////////////////////////////////////////////////////////////////////
  // Test method call functionality
  //////////////////////////////////////////////////////////////////////
, 'A method call' : {
    // Test invalid internal URI to send method call to
    'with an invalid internal URI' : {
      topic: function () {
        var client = new Client({ host: 'localhost', port: 9999, path: '/'}, false)
        client.methodCall('getArray', null, this.callback)
      }
    , 'contains the error' : function (error, value) {
        assert.isObject(error)
      }
    }
  , 'with a string URI for options' : {
      topic: function () {
        var that = this
        // Basic http server that sends a chunked XML response
        http.createServer(function (request, response) {
            response.writeHead(200, {'Content-Type': 'text/xml'})
            var data = '<?xml version="2.0" encoding="UTF-8"?>'
              + '<methodResponse>'
              + '<params>'
              + '<param><value><string>more.listMethods</string></value></param>'
              + '</params>'
              + '</methodResponse>'
            response.write(data)
            response.end()
        }).listen(9090, 'localhost', function() {
          var client = new Client('http://localhost:9090', false)
          client.methodCall('listMethods', null, that.callback)
        })
      }
    , 'contains the string' : function (error, value) {
        assert.isNull(error)
        assert.deepEqual(value, 'more.listMethods')
      }
    }
  , 'with no host specified' : {
      topic: function () {
        var that = this
        // Basic http server that sends a chunked XML response
        http.createServer(function (request, response) {
            response.writeHead(200, {'Content-Type': 'text/xml'})
            var data = '<?xml version="2.0" encoding="UTF-8"?>'
              + '<methodResponse>'
              + '<params>'
              + '<param><value><string>system.listMethods</string></value></param>'
              + '</params>'
              + '</methodResponse>'
            response.write(data)
            response.end()
        }).listen(9091, 'localhost', function() {
          var client = new Client({ port: 9091, path: '/'}, false)
          client.methodCall('listMethods', null, that.callback)
        })
      }
    , 'contains the string' : function (error, value) {
        assert.isNull(error)
        assert.deepEqual(value, 'system.listMethods')
      }
    }
  , 'with a chunked response' : {
      topic: function () {
        var that = this
        // Basic http server that sends a chunked XML response
        http.createServer(function (request, response) {
          response.writeHead(200, {'Content-Type': 'text/xml'})
          var chunk1 = '<?xml version="2.0" encoding="UTF-8"?>'
            + '<methodResponse>'
            + '<params>'
            + '<param><value><array><data>'
            + '<value><string>system.listMethods</string></value>'
            + '<value><string>system.methodSignature</string></value>'
          var chunk2 = '<value><string>xmlrpc_dialect</string></value>'
            + '</data></array></value></param>'
            + '</params>'
            + '</methodResponse>'
          response.write(chunk1)
          response.write(chunk2)
          response.end()
        }).listen(9092, 'localhost', function() {
          var client = new Client({ host: 'localhost', port: 9092, path: '/'}, false)
          client.methodCall('listMethods', null, that.callback)
        })
      }
    , 'contains the array' : function (error, value) {
        assert.isNull(error)
        assert.deepEqual(value, ['system.listMethods', 'system.methodSignature', 'xmlrpc_dialect'])
      }
    }
  , 'with gzip encoded request and response (and utf8)' : {
      topic: function () {
        var that = this;
        http.createServer(function (request, response) {
          response.writeHead(200, {'Content-Type': 'text/xml', 'Content-Encoding': 'gzip'});
          var data = '<?xml version="2.0" encoding="UTF-8"?>'
            + '<methodResponse>'
            + '<params>'
            + '<param><value><string>here is mr. Snowman: ☃</string></value></param>'
            + '</params>'
            + '</methodResponse>';
          zlib.gzip(data, function (err, buffer) {
            response.write(buffer);
            response.end();
          });
        }).listen(9190, 'localhost', function() {
          var client = new Client({
            host: 'localhost',
            port: 9190,
            gzip: 'both'
          }, false);
          client.methodCall('listMethods', null, that.callback);
        })
      }
    , 'contains the correct string' : function (error, value) {
        assert.isNull(error);
        assert.deepEqual(value, 'here is mr. Snowman: ☃');
      }
    }
  , 'with a utf-8 encoding' : {
      topic: function () {
        var that = this
        http.createServer(function (request, response) {
            response.writeHead(200, {'Content-Type': 'text/xml'})
            var data = '<?xml version="2.0" encoding="UTF-8"?>'
              + '<methodResponse>'
              + '<params>'
              + '<param><value><string>here is mr. Snowman: ☃</string></value></param>'
              + '</params>'
              + '</methodResponse>'
            response.write(data)
            response.end()
        }).listen(9093, 'localhost', function() {
          var client = new Client('http://localhost:9093', false)
          client.methodCall('listMethods', null, that.callback)
        })
      }
    , 'contains the correct string' : function (error, value) {
        assert.isNull(error)
        assert.deepEqual(value, 'here is mr. Snowman: ☃')
      }
    }
  , 'with a ISO-8859-1 encoding' : {
      topic: function () {
        var that = this
        http.createServer(function (request, response) {
          response.writeHead(200, {'Content-Type': 'text/xml'})
          // To prevent including a npm package that needs to compile (iconv):
          // The following iso 8859-1 text below in hex
          //   <?xml version="1.0" encoding="ISO-8859-1"?>
          //   <methodResponse>
          //   <params>
          //   <param><value><string>äè12</string></value></param>
          //   </params>
          //   </methodResponse>
          var hex = '3c3f786d6c2076657273696f6e3d22312e302220656e636'
            + 'f64696e673d2249534f2d383835392d31223f3e3c6d6574686f64'
            + '526573706f6e73653e3c706172616d733e3c706172616d3e3c766'
            + '16c75653e3c737472696e673ee4e831323c2f737472696e673e3c'
            + '2f76616c75653e3c2f706172616d3e3c2f706172616d733e3c2f6'
            + 'd6574686f64526573706f6e73653e'
          var hexData = new Buffer(hex, 'hex')
          response.write(hexData)
          response.end()
        }).listen(9094, 'localhost', function() {
          var client = new Client({ host: 'localhost', port: 9094, path: '/', responseEncoding : 'binary'}, false)
          client.methodCall('listMethods', null, that.callback)
        })
      }
    , 'contains the correct string' : function (error, value) {
        assert.isNull(error)
        // assert.deepEqual(value, 'äè12')
        // TODO: Fix client to support other encodings.
      }
    }
  , 'with a multi-byte character in request' : {
      topic: function () {
        var that = this
          , requestBody = ''
        http.createServer(function (request, response) {
          request.setEncoding('utf8')
          request.on('data', function (chunk) {
            requestBody += chunk
          })
          request.on('end', function () {
            response.writeHead(200, {'Content-Type': 'text/xml'})
            var data = '<?xml version="2.0" encoding="UTF-8"?>'
              + '<methodResponse>'
              + '<params>'
              + '<param><value><string>ok</string></value></param>'
              + '</params>'
              + '</methodResponse>'
            response.write(data)
            response.end()
          })
        }).listen(9095, 'localhost', function() {
          var client = new Client({ host: 'localhost', port: 9095, path: '/'}, false)
          client.methodCall('multiByte', ['ö'], function (error) {
            that.callback(error, requestBody)
          })
        })
      }
    , 'contains full request' : function (error, value) {
        var data = '<?xml version="1.0"?>'
          + '<methodCall>'
          + '<methodName>multiByte</methodName>'
          + '<params><param><value><string>ö</string></value></param></params>'
          + '</methodCall>'
        assert.isNull(error)
        assert.deepEqual(value, data)
      }
    }
  , 'with an unknown request' : {
      topic: function () {
        var that = this
        http.createServer(function(request, response) {
          response.writeHead(404)
          response.end()
        }).listen(9099, 'localhost', function() {
          var client = new Client({ host: 'localhost', port: 9099, path: '/'}, false)
          client.methodCall('unknown', null, function (error) {that.callback(error)})
        })
      }
    , 'return NotFound Error' : function (error, value) {
        assert.isObject(error)
      }
    }
  , 'with cookies in response' : {
      topic: function (){
        var that = this
        var invokeCount = 0
        http.createServer(function(request, response) {
          response.writeHead(200, {'Content-Type': 'text/xml', 'Set-Cookie': 'a=b'})
          response.write(VALID_RESPONSE)
          response.end()
          invokeCount++
          if (invokeCount == 2) {
            that.callback(undefined, request.headers['cookie'])
          }
        }).listen(9096, 'localhost', function() {
            var client = new Client({ host: 'localhost', port: 9096, path: '/', cookies: true}, false);
            function cbIfError(err, result) {
              if (err) that.callback(err, result);
            }
            client.methodCall('1', null, function(err, result) {
              cbIfError(err, result);
              client.methodCall('2', null, cbIfError);
            })
          })

      },
      'sends them back to the server' : function(error, value) {
        assert.isNull(error, 'Error was received but not expected');
        assert.equal(value, 'a=b');
      }
    }
  , 'that responds with a malformed xml': {
      topic: function () {
        var that = this
        http.createServer(function(request, response) {
          response.writeHead(500, {'Content-Type': 'text/html'})
          response.write(BROKEN_XML)
          response.end()
        }).listen(9097, 'localhost', function () {
          var client = new Client({ host: 'localhost', port: 9097, path: '/' }, false)
          client.methodCall('broken', null, that.callback)
        })
      }
    , 'returns an error': function (error, value) {
        assert.instanceOf(error, Error)
        assert.match(error.message, /^Unexpected end/)
      }
    , 'returns the request object with the error': function (error, value) {
        assert.instanceOf(error, Error)
        //assert.isObject(error.req)
        //assert.isObject(error.req.connection)
        //assert.isString(error.req._header)
      }
    , 'returns the response object with the error': function (error, value) {
        assert.instanceOf(error, Error)
        //assert.isObject(error.res)
        //assert.strictEqual(error.res.statusCode, 500)
      }
    , 'returns the body with the error': function (error, value) {
        //assert.strictEqual(error.body, BROKEN_XML.toString())
      }
    }
  }
}).export(module)


lib/xmlrpc.js
var Client = require('./client')
  , Server = require('./server')
  , CustomType = require('./customtype')
  , dateFormatter = require('./date_formatter')

var xmlrpc = exports

/**
 * Creates an XML-RPC client.
 *
 * @param {object} options - server options to make the HTTP request to
 * @param {string} options.host
 * @param {number} options.port
 * @param {string} options.url
 * @param {boolean} [options.cookies]
 * @param {number} [options.keepALive] keepAlive Miliseconds, 0 to disable (default).
 * @param {string} [options.gzip] GZip support, false default. could be 'response' or 'both'.
 * @param {string} [options.userAgent] Custom user agent, optional.
 * @return {Client}
 * @see Client
 */
xmlrpc.createClient = function(options) {
  return new Client(options, false)
}

/**
 * Creates an XML-RPC client that makes calls using HTTPS.
 *
 * @param {object} options - server options to make the HTTP request to
 * @param {string} options.host
 * @param {number} options.port
 * @param {string} options.url
 * @param {boolean} [options.cookies]
 * @param {number} [options.keepALive] keepAlive Miliseconds, 0 to disable (default).
 * @param {string} [options.gzip] GZip support, false default. could be 'response' or 'both'.
 * @param {string} [options.userAgent] Custom user agent, optional.
 * @return {Client}
 * @see Client
 */
xmlrpc.createSecureClient = function(options) {
  return new Client(options, true)
}

/**
 * Creates an XML-RPC server.
 *
 * @param {Object}options - the HTTP server options
 *   - {String} host
 *   - {Number} port
 * @return {Server}
 * @see Server
 */
xmlrpc.createServer = function(options, callback) {
  return new Server(options, false, callback)
}

/**
 * Creates an XML-RPC server that uses HTTPS.
 *
 * @param {Object}options - the HTTP server options
 *   - {String} host
 *   - {Number} port
 * @return {Server}
 * @see Server
 */
xmlrpc.createSecureServer = function(options, callback) {
  return new Server(options, true, callback)
}

xmlrpc.CustomType = CustomType
xmlrpc.dateFormatter = dateFormatter


lib/client.js
var url          = require('url')
  , zlib         = require('zlib')
  , Agent        = require('agentkeepalive')
  , superagent   = require('superagent').agent()
  , Serializer   = require('./serializer')
  , Deserializer = require('./deserializer')
  , Cookies      = require('./cookies');

/**
 * Creates a Client object for making XML-RPC method calls.
 *
 * @constructor
 * @param {Object|string} options - Server options to make the HTTP request to.
 *                                  Either a URI string
 *                                  (e.g. 'http://localhost:9090') or an object
 *                                  with fields:
 *   {string} options.host              - (optional)
 *   {number} options.port
 *   {string} options.url               - (optional) - may be used instead of host/port pair
 *   {boolean} options.cookies          - (optional) - if true then cookies returned by server will be stored and sent back on the next calls.
 *   {number} options.keepAlive         - (optional) - Keep Alive Miliseconds, set to 0 to disable.
 *   {string} options.gzip              - (optional) - GZip compression, default false. 'response' or 'both', both is sending and receiving (kind of forcing)
 *   {string} options.userAgent         - (optional) - Custom User-Agent.
 *                                  Also it will be possible to access/manipulate cookies via #setCookie/#getCookie methods
 * @param {boolean} isSecure      - True if using https for making calls,
 *                                  otherwise false.
 * @return {Client}
 */
function Client(options, isSecure) {

  // Invokes with new if called without
  if (false === (this instanceof Client)) {
    return new Client(options, isSecure)
  }

  // If a string URI is passed in, converts to URI fields
  if (typeof options === 'string') {
    options = url.parse(options)
    options.host = options.hostname
    options.path = options.pathname
  }

  if (typeof options.url !== 'undefined') {
    var parsedUrl = url.parse(options.url);
    options.host = parsedUrl.hostname;
    options.path = parsedUrl.pathname;
    options.port = parsedUrl.port;
  }

  if (! options.host) {
    options.host = 'localhost';
  }
  if (! options.port) {
    options.port = (isSecure ? 443 : 80);
  }
  if (! options.path) {
    options.path = '/';
  }
  if (options.path && options.path.charAt(0) !== '/') {
    options.path = '/' + options.path;
  }
  options.keepAlive = options.keepAlive || 0;
  options.gzip      = options.gzip || false;

  options.url       = (isSecure ? 'https://' : 'http://') + options.host + ':';
  options.url      += options.port + options.path;

  // Set the HTTP request headers
  var headers = {
    'User-Agent'     : options.userAgent || 'NodeJS XML-RPC Client/' + require('./../package').version
  , 'Content-Type'   : 'text/xml'
  , 'Accept'         : 'text/xml'
  , 'Accept-Charset' : 'UTF8'
  , 'Connection'     : 'Keep-Alive'
};

  // Compression Header.
  if (options.gzip) {
    headers['Accept-Encoding'] = 'gzip';
  }
  if (options.gzip === 'both') {
    // Also encode request!
    headers['Content-Encoding'] = 'gzip';
  }

  options.headers = options.headers || {};

  if (options.headers.Authorization == null &&
      options.basic_auth != null &&
      options.basic_auth.user != null &&
      options.basic_auth.pass != null)
  {
    var auth = options.basic_auth.user + ':' + options.basic_auth.pass;
    options.headers['Authorization'] = 'Basic ' + new Buffer(auth).toString('base64');
  }

  for (var attribute in headers) {
    if (options.headers[attribute] === undefined) {
      options.headers[attribute] = headers[attribute];
    }
  }

  options.method = 'POST';
  this.options = options;

  this.isSecure = isSecure;
  this.headersProcessors = {
    processors: [],
    composeRequest: function(headers) {
      this.processors.forEach(function(p) {p.composeRequest(headers);});
    },
    parseResponse: function(headers) {
      this.processors.forEach(function(p) {p.parseResponse(headers);});
    }
  };
  if (options.cookies) {
    this.cookies = new Cookies();
    this.headersProcessors.processors.unshift(this.cookies);
  }

  var agentOptions = {
    maxSockets: 100,
    maxFreeSockets: 10,
    timeout: 3000
  };

  if (options.keepAlive > 0) {
    agentOptions.timeout = options.keepAlive + 10000;
    agentOptions.keepAliveTimeout = options.keepAlive;
  }
  this.agent = new Agent(agentOptions);
}

/**
 * Makes an XML-RPC call to the server specified by the constructor's options.
 *
 * @param {String} method     - The method name.
 * @param {Array} params      - Params to send in the call.
 * @param {Function} callback - function(error, value) { ... }
 *   - {Object|null} error    - Any errors when making the call, otherwise null.
 *   - {mixed} value          - The value returned in the method response.
 */
Client.prototype.methodCall = function methodCall(method, params, callback) {
  var self      = this;
  var options   = this.options;
  var xml       = Serializer.serializeMethodCall(method, params, options.encoding);

  this.headersProcessors.composeRequest(options.headers);

  var request = superagent.post(options.url);
  request.set(options.headers);
  request.agent(this.agent);

  if (options.gzip && options.gzip === 'both') {
    zlib.gzip(xml, function (err, buffer) {
      request.set('Content-Length', buffer.length);
      request.send(buffer);
    });
  } else {
    request.set('Content-Length', Buffer.byteLength(xml, 'utf8'));
    request.send(xml);
  }

  var ended = false;
  request.on('error', function (err) {
    if (! ended) {
      ended = true;
      callback(err);
    }
  });
  request.on('end', function () {
    ended = true;
    if (this.err) {
      return callback(this.err);
    }
    if (this.res.statusCode == 404) {
      return callback(new Error('Not Found'));
    }
    self.headersProcessors.parseResponse(this.res.headers);
  });

  var deserializer = new Deserializer(options.responseEncoding);
  deserializer.deserializeMethodResponse(request, function(err, result) {
    if (ended) { return callback(err, result); }
    request.on('end', function() {
      callback(err, result);
    });
  });
}

/**
 * Gets the cookie value by its name. The latest value received from servr with 'Set-Cookie' header is returned
 * Note that method throws an error if cookies were not turned on during client creation (see comments for constructor)
 *
 * @param {String} name name of the cookie to be obtained or changed
 * @return {*} cookie's value
 */
Client.prototype.getCookie = function getCookie(name) {
  if (!this.cookies) {
    throw 'Cookies support is not turned on for this client instance';
  }
  return this.cookies.get(name);
}

/**
 * Sets the cookie value by its name. The cookie will be sent to the server during the next xml-rpc call.
 * The method returns client itself, so it is possible to chain calls like the following:
 *
 * <code>
 *   client.cookie('login', 'alex').cookie('password', '123');
 * </code>
 *
 * Note that method throws an error if cookies were not turned on during client creation (see comments for constructor)
 *
 * @param {String} name name of the cookie to be changed
 * @param {String} value value to be set.
 * @return {*} client object itself
 */
Client.prototype.setCookie = function setCookie(name, value) {
  if (!this.cookies) {
    throw 'Cookies support is not turned on for this client instance';
  }
  this.cookies.set(name, value);
  return this;
}

module.exports = Client
